### ISSUE TITLE
Nested ARRAY_AGG generates invalid code

### ISSUE THREAD

**Issue by roddyaj** [jOOQ/jOOQ#13454]
### Expected behavior
View with ARRAY_AGG of ARRAY_AGG generates code that compiles.

### Actual behavior
View with ARRAY_AGG of ARRAY_AGG generates code that doesn't compile.  The crux of the matter can be seen in the table class in this line:
`public final TableField<MyViewRecord, Object[][]> IDS = createField(DSL.name("IDS"), SQLDataType.OTHER.getArrayDataType(), this, "");`
It cannot convert from TableField<MyViewRecord,Object[]> to TableField<MyViewRecord,Object[][]>.
I'm not sure if the intent is to use Object[] or Object[][], but if I replace all occurrences of [][] with [] in the table and record classes, everything works fine for me.
This issue came up for me when upgrading from 3.14.7 to 3.16.5.

### Steps to reproduce the problem

CREATE TABLE MY_TABLE (ID INT, NAME VARCHAR);
CREATE VIEW MY_VIEW AS (
	SELECT ARRAY_AGG(IDS) AS IDS FROM
		(SELECT ARRAY_AGG(ID) AS IDS FROM MY_TABLE GROUP BY NAME)
);
Configuration configuration = new Configuration()
	.withLogging(Logging.WARN)
	.withJdbc(new Jdbc()
		.withDriver("org.h2.Driver")
		.withUrl("jdbc:h2:mem:myproject;DB_CLOSE_DELAY=-1")
		.withUser("my_user")
		.withPassword(""))
	.withGenerator(new Generator()
		.withGenerate(new Generate()
			.withJavaTimeTypes(false)
			.withSpatialTypes(false))
		.withDatabase(new Database()
			.withName("org.jooq.meta.h2.H2Database")
			.withIncludes(".*")
			.withExcludes(".*flyway.*|.*Intermediary.*|.*_IDX_COLUMN")
			.withIncludeExcludeColumns(true)
			.withInputSchema("PUBLIC")
			.withCustomTypes(
				new CustomType().withName("Geometry").withType("org.locationtech.jts.geom.Geometry").withBinding("com.myproject.jooq.geometry.JTSGeometryBinding"))
			.withForcedTypes(
				new ForcedType().withName("Geometry").withIncludeTypes("(geometry|GEOMETRY)"),
				new ForcedType().withUserType("com.fasterxml.jackson.databind.node.ArrayNode").withBinding("com.myproject.jooq.json.ArrayNodeBinding").withIncludeExpression(".*JSON_ARRAY").withIncludeTypes(".*"),
				new ForcedType().withUserType("com.fasterxml.jackson.databind.node.ObjectNode").withBinding("com.myproject.jooq.json.ObjectNodeBinding").withIncludeExpression(".*JSON").withIncludeTypes(".*")))
		.withTarget(new Target()
			.withPackageName("com.myproject.jooq.codegen")
			.withDirectory("./src/main/java")));

### Versions

- jOOQ:
- Java:
- Database (include vendor):
- OS:
- JDBC Driver (include name if inofficial driver):

**Comment by lukaseder** [comment]
Thanks a lot for your report. jOOQ does not yet support PostgreSQL's multi dimensional arrays (see https://github.com/jOOQ/jOOQ/issues/252), so if this had worked in the past, it worked by accident.

However, perhaps it's time to finally support the feature. I'll have another look later this week.

**Comment by roddyaj** [comment]
Thanks!  Note that I'm using H2, not Postgres, if that's relevant.  Yes this was working fine in 3.14.7 and probably other earlier versions - I only have to cast the array objects appropriately in my client code :)

**Comment by lukaseder** [comment]
> Thanks! Note that I'm using H2, not Postgres, if that's relevant.

H2 2.0 implemented support for standard SQL arrays, so they should work more or less the same way

> Yes this was working fine in 3.14.7 and probably other earlier versions - I only have to cast the array objects appropriately in my client code

I'm guessing you were using H2 1.4 at the time, because jOOQ 3.14 didn't support H2 2.0 yet. Then I'm doubly surprised that it worked, given H2 1.4's very own interpretation of array types.

**Comment by roddyaj** [comment]
Yes I was using H2 1.4.
With jOOQ 3.14 / H2 1.4 I got an Object[] of Object[] (the inner ones being numbers in this case).
With jOOQ 3.16 / H2 2.1 I got an Object[] of JdbcArray (with slight modification of the generated code that I mentioned above).

**Comment by lukaseder** [comment]
I think that implementing https://github.com/jOOQ/jOOQ/issues/6454 should fix this.

**Comment by lukaseder** [comment]
Yep, #6454 will fix this here. Will implement and backport to 3.16

**Comment by lukaseder** [comment]
HSQLDB doesn't support nesting arrays this way. It produces:

> SQL Error [42534]: column of LOB or ARRAY type cannot be used in operation

**Comment by lukaseder** [comment]
Neither does CockroachDB: https://github.com/cockroachdb/cockroach/issues/32552?version=v21.1

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.17.0 and 3.16.7 (https://github.com/jOOQ/jOOQ/issues/13532)

**Comment by katzyn** [comment]
@lukaseder
With JDBC driver of H2 `ResultSet.getObject(column, Integer[][].class)` can be used to read values of `INTEGER ARRAY ARRAY` data type as Java arrays directly without excessive conversions between various representations.

Simple `getObject(column)` returns an implementation of `java.sql.Array` because such behavior is required by appendix B.3 JDBC Types Mapped to Java Object Types of JDBC Specification.

**Comment by lukaseder** [comment]
Thanks, @katzyn. #6454 was a reasonable feature request anyway, and given that it transparently fixed this issue here, I'll leave things as they are.
